secant/Features/BackupFlow/RecoveryPhraseValidation.swift

```diff
@@ -0,0 +1,231 @@
+            return .incomplete
+        }
+
+        return .complete
+    }
+
+    var isValid: Bool {
+        self.resultingPhrase == phrase.words
+    }
+}
+
+extension RecoveryPhraseValidationViewStore {
+    func bindingForRoute(_ route: RecoveryPhraseValidationState.Route) -> Binding<Bool> {
+        self.binding(
+            get: { $0.route == route },
+            send: { isActive in
+                return .updateRoute(isActive ? route : nil)
+            }
+        )
+    }
+}
+
+extension RecoveryPhraseValidationState {
+    /// creates an initial `RecoveryPhraseValidationState` with no completions and random missing indices.
+    /// - Note: Use this function to create a random validation puzzle for a given phrase.
+    static func random(phrase: RecoveryPhrase) -> Self {
+        let missingIndices = Self.randomIndices()
+        let missingWordChipKind = phrase.words(fromMissingIndices: missingIndices)
+        return RecoveryPhraseValidationState(
+            phrase: phrase,
+            missingIndices: missingIndices,
+            missingWordChips: missingWordChipKind,
+            validationWords: []
+        )
+    }
+}
+
+extension RecoveryPhraseValidationState {
+    /// Given an array of RecoveryPhraseStepCompletion, missing indices, original phrase and the number of groups it was split into,
+    /// assembly the resulting phrase. This comes up with the "proposed solution" for the recovery phrase validation challenge.
+    /// - returns:an array of String containing the recovery phrase words ordered by the original phrase order.
+    var resultingPhrase: [String] {
+        assert(missingIndices.count == validationWords.count)
+        assert(validationWords.count == Self.phraseChunks)
+
+        var words = phrase.words
+        let groupLength = words.count / Self.phraseChunks
+        // iterate based on the completions the user did on the UI
+        for wordCompletion in validationWords {
+            // figure out which phrase group (chunk) this completion belongs to
+            let i = wordCompletion.groupIndex
+            // validate that's the right number
+            assert(i < Self.phraseChunks)
+            // get the missing index that the user did this completion for on the given group
+            let missingIndex = missingIndices[i]
+            // figure out what this means in terms of the whole recovery phrase
+            let concreteIndex = i * groupLength + missingIndex
+            assert(concreteIndex < words.count)
+            // replace the word on the copy of the original phrase with the completion the user did
+            words[concreteIndex] = wordCompletion.word
+        }
+
+        return words
+    }
+
+    /// drives the state machine represented on this Enum by the action of applying a chip into a group of words containing an empty slot that has to be completed
+    func apply(chip: PhraseChip.Kind, into group: Int) -> Self {
```

Comment:
See longer comment about removing this function and having the logic directly in the reducer, but more succinct.

Reply:
done